edb/schema/utils.py

```diff
@@ -1167,6 +1167,128 @@ def get_non_overlapping_union(
         return frozenset(all_objects), True
 
 
+def get_type_expr_non_overlapping_union(
+    type: s_types.Type,
+    schema: s_schema.Schema,
+) -> Tuple[FrozenSet[s_types.Type], bool]:
+    """Get a non-overlapping set of the type's descendants"""
+
+    from edb.schema import types as s_types
+
+    expanded_types = expand_type_expr_descendants(type, schema)
+
+    # filter out subclasses
+    expanded_types = {
+        type
+        for type in expanded_types
+        if not any(
+            type is not other and type.issubclass(schema, other)
+            for other in expanded_types
+        )
+    }
```

Comment:
Is this needed? I guess it helps us not always require `union_is_exhaustive`?

Reply:
That's right. It probably doesn't matter from a performance point of view, but it does make reading the sql easier too.